don't fail trying to remove children with absent parents

evgeni opened the pull request:
Fixes: #1301

evgeni committed:
don't fail trying to remove children with absent parents

Fixes: #1301

evgeni commented:
this certainly deserves a test, but wanted to show what I have for now

mdellweg commented on plugins/module_utils/foreman_helper.py:
```diff
@@ -1315,6 +1315,8 @@ def __init__(self, **kwargs):
             if 'resouce_type' not in self.foreman_spec['parent']:
                 self.foreman_spec['parent']['resource_type'] = self.foreman_spec['entity']['resource_type']
+            if 'failsafe' not in self.foreman_spec['parent']:
+                self.foreman_spec['parent']['failsafe'] = True
```
This prevents the lookup failure...

mdellweg commented on plugins/module_utils/foreman_helper.py:
```diff
@@ -1424,10 +1426,14 @@ def __init__(self, **kwargs):
+                and 'parent' in self.foreman_params and self.lookup_entity('parent') is None):
+            if self.desired_absent:
+                # Parent does not exist so just exit here
+                return None
```
... this prevents that the missing parent is a problem ...

mdellweg commented on plugins/module_utils/foreman_helper.py:
```diff
@@ -1424,10 +1426,14 @@ def __init__(self, **kwargs):
+                # Parent does not exist so just exit here
+                return None
+            else:
+                self.fail_json(msg="Couldn't find parent '{0}' for '{1}'.".format(parent_name, self.foreman_params['name']))
```
... and this is the postponed failure.

Would it be possible to make the failsafe flag dependent on desired_absent in the first place?

evgeni replied:
I thought about that, but `desired_absent` is only present in `ForemanEntityAnsibleModule`, but the `parent` logic is in `ForemanStatelessEntityAnsibleModule`, which doesn't have `state` XD

mdellweg replied:
Sounds like a bigger mess.

evgeni replied:
Yes.

mdellweg commented on plugins/module_utils/foreman_helper.py:
```diff
@@ -1424,10 +1426,14 @@ def __init__(self, **kwargs):
     def run(self, **kwargs):
         """ lookup entities, ensure entity, remove sensitive data, manage parameters.
         """
+        parent_name = self.foreman_params.get('parent')
```
This is only used in line 1436. Maybe move it there.

evgeni replied:
No, as line 1430/1431 has `self.lookup_entity('parent')` which overwrites this with `None` ;)

mdellweg reviewed (commented):
Looks good so far. I suggested another approach, but that may or may not create an even bigger mess.

evgeni committed:
update location tests to test for non-existant-parent delete

evgeni committed:
update hostgroup tests to test for non-existant-parent delete

evgeni committed:
ensure location tests can fail

before this, if the call to the location module failed, but
expected_error was not set, the failure was just swallowed

mdellweg commented on tests/test_playbooks/tasks/location.yml:
```diff
@@ -21,6 +21,12 @@
+    that:
+      - result is succeeded
+  when:
+    - expected_error is not defined or not expected_error
```
Can we use `expected_error is defined` as a condition for ignore_errors?

evgeni replied:
maybe? I'll try

evgeni replied:
YES!
